show error when running dockershim

MHBauer opened the pull request:
The error returned when running the dockershim is not checked.
When running dockerless, we should show the existing error.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
All errors when running dockershim are silently ignored. We should surface errors, especially when they tell you exactly what's wrong.

**Which issue(s) this PR fixes**:
Found while looking at #91657

**Special notes for your reviewer**:
Run `GOFLAGS='-tags=dockerless' make test-e2e-node PARALLELISM=1 TEST_ARGS='--kubelet-flags=--fail-swap-on=false' FOCUS="Legacy:Docker" SKIP="" RUNTIME=docker`

Then check kublet logs for something like:
```
Jun 04 12:39:08 mhb-T470p kubelet[23994]: F0604 12:39:08.482671   23994 server.go:265] failed to run Kubelet: trying to use docker runtime when Kubelet was compiled without docker support
```

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
In theory, related to:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20200205-build-kubelet-without-docker.md
```

MHBauer commented:
/cc @mattjmcnaughton @dims 
/assign @derekwaynecarr

mattjmcnaughton reviewed (approved):
/retest

You are 100% correct, great catch!

/lgtm

As an added bonus - can you look into how easy it would be to add a unit test that ensures that any errors raised in running the dockershim result in `PreInitRuntimeService` returning an error? (I.e. a regression test for this bug which was introduced during the `dockerless` refactor). Its also fine w/ me if that comes as a separate pr.

cc @dims @derekwaynecarr if y'all can approve this soon, that'd be great. The `dockerless` change wasn't intended to introduce any behavior changes, but until we merge this diff, that assumption isn't true. Thanks!

MHBauer commented:
@mattjmcnaughton amended with a set of very basic tests. I can follow up with some more complicated stuff if desired.

mattjmcnaughton commented on pkg/kubelet/kublet_dockerless_test.go:
```diff
@@ -0,0 +1,17 @@
+// +build dockerless
```
Hmmmm, I'm curious for the thoughts of others on whether it makes sense to have a `kubelet_dockerless_test.go`...

If we do decide to keep this file, can you fix the spelling in the name?

Also, we need to run `hack/update-bazel.sh`.

MHBauer replied:
I must make that typo 50% of the time.

Update and do both of those now, thanks.

mattjmcnaughton reviewed (commented):
Thanks for the test coverage addition @MHBauer !

Would love to hear from others on the test coverage. Regardless of what we decide on the tests, I support returning the error if it isn't `nil`.

MHBauer commented:
I ran `hack/update-bazel.sh` but there wasn't anything extra to commit.

I'm getting the same (non-error?) message as in https://github.com/kubernetes/kubernetes/issues/79499#issuecomment-614455725.

mattjmcnaughton commented:
@MHBauer would you be open to breaking up the diff fixing the bug and the diff adding the unit tests into two separate diffs?

I'd like us to merge the one fixing the bug asap - the test component seems to have some funky stuff going on, which I'm sure we'll figure it, but I'd prefer to not block the bug fix on it :)

Thanks again for catching this :)

MHBauer commented:
separate PRs or separate commits? I think PRs is what you mean by diff.

MHBauer committed:
show error when running dockershim

The error returned when running the dockershim is not checked.
When running dockerless, we should show the existing error.

matthyx commented:
/uncc

mattjmcnaughton commented:
On Tue, Jun 09, 2020 at 11:25:24PM -0700, Morgan Bauer wrote:
> separate PRs or separate commits? I think PRs is what you mean by diff.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/kubernetes/kubernetes/pull/91789#issuecomment-641753889

Yup, I was thinking separate PRs, so that the weirdness w/ the testing doesn't
block the bug fix.

MHBauer commented:
Only the error in this PR
Looks like the failing jobs are timing out? 
/retest

derekwaynecarr commented:
/approve
/lgtm

mattjmcnaughton commented:
Excellent, thanks @MHBauer !
